src/status_im/protocol/web3/inbox.cljs

```diff
@@ -1,67 +1,79 @@
+    (-> s
+        js/JSON.parse
+        (js->clj :keywordize-keys true))
+    (catch :default e
+      {:error (.-message e)})))
 
-;; TODO(oskarth): Rewrite callback-heavy code with CSP and/or coeffects
-;; TODO(oskarth): Memoize addPeer and markTrusted, similar to keys/get-sym-key
-;; TODO(oskarth): Actually deal with errors, all in same cb - outside scope of this
-(defn request-messages! [web3 {:keys [enode topic password]
-                               :or {enode    default-enode
-                                    password inbox-password
-                                    topic    default-topic}} callback]
-  (status/add-peer
-   enode
-   (fn [res]
-     (log/info "offline inbox: add peer" enode res)
-     (let [args {:jsonrpc "2.0"
-                 :id      1
-                 :method  "shh_markTrustedPeer"
-                 :params  [enode]}
-           payload (.stringify js/JSON (clj->js args))]
-       (log/info "offline inbox: mark-trusted-peer request")
-       (status/call-web3
-        payload
-        (fn [res2]
-          (log/info "offline inbox: mark-trusted-peer response" enode res2)
-          (keys/get-sym-key web3 password
-                            (fn [sym-key-id]
-                              (log/info "offline inbox: sym-key-id" sym-key-id)
-                              (let [args {:jsonrpc "2.0"
-                                          :id      2
-                                          :method  "shh_requestMessages"
-                                          ;; NOTE: "from" and "to" parameters omitted here
-                                          ;; by default "from" is 24 hours ago and "to" is time now
-                                          :params  [{:mailServerPeer enode
-                                                     :topic          topic
-                                                     :symKeyID       sym-key-id}]}
-                                    payload (.stringify js/JSON (clj->js args))]
-                                (log/info "offline inbox: request-messages request")
-                                (log/info "offline inbox: request-messages args" (pr-str args))
-                                (log/info "offline inbox: request-messages payload" (pr-str payload))
+(defn- response-handler [error-fn success-fns]
+  (fn [response]
+    (let [{:keys [error result]} (parse-json response)]
+      ;; NOTE(dmitryn): AddPeer() may return {"error": ""}
+      ;; assuming empty error is a success response
+      (if (or (nil? error)
+              (string/blank? error))
+        (doseq [f success-fns]
+          (f result))
+        (error-fn error)))))
 
-                                (status/call-web3 payload callback))))))))))
+(defn add-peer [enode success-fn error-fn]
+  (if (@peers enode)
+    (success-fn true)
+    (status/add-peer enode
+                     (response-handler error-fn
+                                       [#(swap! peers conj enode) success-fn]))))
```

Comment:
How come this is a collection of functions as opposed to just one?

Reply:
hm, what's wrong with fns collection? do you see other way to run custom fn (which does swap! in this case) on success event in `response-handler`?

Reply:
Nothing wrong per se? I'd probably put [f1 f2] into f3 to simplify interface, but not a big deal

Reply:
I revisited the implementation and changed my mind, wrapping to fn1 and fn2 to f3 really simplifies interface so i changed that too, thanks for advice.